internal/registry/registry_cache_test.go

```diff
@@ -0,0 +1,227 @@
+			})
+		})
+
+		it("creates a RegistryCache", func() {
+			registryCache, err := NewRegistryCache(logger, tmpDir, registryFixture)
+			h.AssertNil(t, err)
+			expectedRoot := filepath.Join(tmpDir, "registry")
+			actualRoot := strings.Split(registryCache.Root, "-")[0]
+			h.AssertEq(t, actualRoot, expectedRoot)
+		})
+	})
+
+	when("#LocateBuildpack", func() {
+		var (
+			registryCache Cache
+		)
+
+		it.Before(func() {
+			registryCache, err = NewRegistryCache(logger, tmpDir, registryFixture)
+			h.AssertNil(t, err)
+		})
+
+		it("locates a buildpack without version", func() {
+			bp, err := registryCache.LocateBuildpack("example/java")
+			h.AssertNil(t, err)
+			h.AssertNotNil(t, bp)
+
+			h.AssertEq(t, bp.Namespace, "example")
+			h.AssertEq(t, bp.Name, "java")
+			h.AssertEq(t, bp.Version, "1.0.0")
+		})
+
+		it("locates a buildpack without version", func() {
+			bp, err := registryCache.LocateBuildpack("example/foo")
+			h.AssertNil(t, err)
+			h.AssertNotNil(t, bp)
+
+			h.AssertEq(t, bp.Namespace, "example")
+			h.AssertEq(t, bp.Name, "foo")
+			h.AssertEq(t, bp.Version, "1.2.0")
+		})
+
+		it("locates a buildpack with version", func() {
+			bp, err := registryCache.LocateBuildpack("example/foo@1.1.0")
+			h.AssertNil(t, err)
+			h.AssertNotNil(t, bp)
+
+			h.AssertEq(t, bp.Namespace, "example")
+			h.AssertEq(t, bp.Name, "foo")
+			h.AssertEq(t, bp.Version, "1.1.0")
+		})
+
+		it("returns error if can't parse buildpack id", func() {
+			_, err := registryCache.LocateBuildpack("quack")
+			h.AssertError(t, err, "parsing buildpacks registry id")
+		})
+
+		it("returns error if buildpack id is empty", func() {
+			_, err := registryCache.LocateBuildpack("example/")
+			h.AssertError(t, err, "empty buildpack name")
+		})
+
+		it("returns error if can't find buildpack with requested id", func() {
+			_, err := registryCache.LocateBuildpack("example/qu")
+			h.AssertError(t, err, "reading entry")
+		})
+
+		it("returns error if can't find buildpack with requested version", func() {
+			_, err := registryCache.LocateBuildpack("example/foo@3.5.6")
+			h.AssertError(t, err, "could not find version")
+		})
+	})
+
+	when("#Refresh", func() {
+		var (
+			registryCache Cache
+		)
+
+		it.Before(func() {
+			registryCache, err = NewRegistryCache(logger, tmpDir, registryFixture)
+			h.AssertNil(t, err)
+		})
+
+		when("registry has new commits", func() {
+			it.Before(func() {
+				err := registryCache.Refresh()
+				h.AssertNil(t, err)
+
+				h.AssertGitHeadEq(t, registryFixture, registryCache.Root)
+
+				r, err := git.PlainOpen(registryFixture)
+				h.AssertNil(t, err)
+
+				w, err := r.Worktree()
+				h.AssertNil(t, err)
+
+				commit, err := w.Commit("second", &git.CommitOptions{
+					Author: &object.Signature{
+						Name:  "John Doe",
+						Email: "john@doe.org",
+						When:  time.Now(),
+					},
+				})
+				h.AssertNil(t, err)
+
+				_, err = r.CommitObject(commit)
+				h.AssertNil(t, err)
+			})
```

Comment:
This is a long `Before` method that applies to a single test. I would prefer to find a way to have this setup take place in the test case. I suspect the case for not doing so is how long it is. I would look to externalize a lot of this into a clearly name helper method (or two, or three) that can be triggered within the test case.

Reply:
I definitely hear that. I simplified it somewhat. The middle chunk is just creating a new commit, which is only done in this test (right now). I'm not sure whether the one use warrants creating an external func for it, what do you think?